turbopack/crates/turbo-tasks/src/graph/graph_store.rs

```diff
@@ -114,3 +114,83 @@ where
         (self.store, VisitedNodes(self.visited))
     }
 }
+
+/// A [`GraphStore`] wrapper that skips nodes that have already been
+/// visited, based on a key extracted from the node.
+///
+/// This is necessary to avoid repeated work when traversing non-tree
+/// graphs (i.e. where a node can have more than one incoming edge).
+#[derive(Debug)]
+pub struct SkipDuplicatesWithKey<StoreImpl, Key, KeyExtractor>
+where
+    StoreImpl: GraphStore,
+    Key: Send + Eq + Hash,
+    KeyExtractor: Send + Fn(&StoreImpl::Node) -> &Key,
+{
+    store: StoreImpl,
+    visited: FxHashSet<Key>,
+    key_extractor: KeyExtractor,
+}
+
+impl<StoreImpl, Key, KeyExtractor> SkipDuplicatesWithKey<StoreImpl, Key, KeyExtractor>
+where
+    StoreImpl: GraphStore,
+    Key: Send + Eq + std::hash::Hash + Clone,
+    KeyExtractor: Send + Fn(&StoreImpl::Node) -> &Key,
+{
+    pub fn new(store: StoreImpl, key_extractor: KeyExtractor) -> Self {
+        Self {
+            store,
+            visited: FxHashSet::default(),
+            key_extractor,
+        }
+    }
+}
+
+impl<StoreImpl, Key, KeyExtractor> GraphStore
+    for SkipDuplicatesWithKey<StoreImpl, Key, KeyExtractor>
+where
+    StoreImpl: GraphStore,
+    StoreImpl::Node: Eq + std::hash::Hash + Clone,
+    Key: Send + Eq + std::hash::Hash + Clone,
+    KeyExtractor: Send + Fn(&StoreImpl::Node) -> &Key,
+{
+    type Node = StoreImpl::Node;
+    type Handle = StoreImpl::Handle;
+
+    fn insert(
+        &mut self,
+        from_handle: Option<Self::Handle>,
+        node: GraphNode<StoreImpl::Node>,
+    ) -> Option<(Self::Handle, &StoreImpl::Node)> {
+        let key = (self.key_extractor)(node.node());
```

Comment:
This is duplicates a lot stuff of `SkipDuplicatesWithKey`, but merging those would be a
- fight with generics
- not sure how performant a `key_extractor: |node| { node }` identity function would be